### PR TITLE
Issue 5737: (SLTS) Improve metrics

### LINKED ISSUE #5737: SLTS - improve metrics
**Problem description**
There are several metrics that are not missing. 

**Problem location**
SLTS

**Suggestions for an improvement**
- Fix old metrics for append only "createWithContent' api, report 'write' instead of 'create' latency.
- New metric - Number active storage threads.
- New metric - Number of links scanned in read.
- New metric - number of metadata entries cached.
- New metric - number of metadata entries misses.

### PR DESCRIPTION
**Pull request by sachin-j-joshi**
**Change log description**  
Improve SLTS metrics by adding more metrics.

**Purpose of the change**  
Fixes #5737

**What the code does**  
Improve SLTS metrics by adding more metrics.

**How to verify it**  
Build should pass.
Visual verification.

### CODE REVIEW

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentStoreMetrics.java [hunk 1]
```diff
@@ -91,13 +91,19 @@ public void close() {
...
+        private final ScheduledExecutorService storageExecutor;
         private final ScheduledFuture<?> reporter;
 
-        public ThreadPool(ScheduledExecutorService executor) {
+        public ThreadPool(ScheduledExecutorService executor, ScheduledExecutorService storageExecutor) {
             this.executor = Preconditions.checkNotNull(executor, "executor");
+            this.storageExecutor = Preconditions.checkNotNull(storageExecutor, "storageExecutor");
             this.queueSize = STATS_LOGGER.createStats(MetricsNames.THREAD_POOL_QUEUE_SIZE);
             this.activeThreads = STATS_LOGGER.createStats(MetricsNames.THREAD_POOL_ACTIVE_THREADS);
+            this.storageQueueSize = STATS_LOGGER.createStats(MetricsNames.STORAGE_THREAD_POOL_QUEUE_SIZE);
```
Comment (andreipaduroiu) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentStoreMetrics.java:105]: You need to close these 2 metrics too.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageMetrics.java [hunk 2]
```diff
@@ -19,6 +20,7 @@
  * Defines all Metrics used by the {@link BaseChunkStorage} class.
  */
 public class ChunkStorageMetrics {
```
Comment (andreipaduroiu) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageMetrics.java:22]: This class should be declared `static` if all it has is static variables and methods.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java [hunk 3]
```diff
@@ -167,6 +171,7 @@ public ChunkedSegmentStorage(int containerId, ChunkStorage chunkStorage, ChunkMe
                 config,
                 (ScheduledExecutorService) executor);
         this.closed = new AtomicBoolean(false);
+        this.reporter = ((ScheduledExecutorService) executor).scheduleAtFixedRate(this::report, 1000, 1000, TimeUnit.MILLISECONDS);
```
Comment (andreipaduroiu) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java:174]: This is dangerous. You may not get a `ScheduledExecutorService`. Please fix your constructor to require that type instead.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/StatsReporter.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.chunklayer;
+
+/**
+ * Contract for components that reports stats.
+ */
+public interface StatsReporter {
+    default void report() {
```
Comment (andreipaduroiu) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/StatsReporter.java:16]: Javadoc .

Also, why do you give it a noop default implementation if you have a single method? If a class does not have anything to report, then maybe it shouldn't implement this interface in the first place.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/StorageMetadataMetrics.java [hunk 5]
```diff
@@ -18,11 +20,18 @@
  * Defines all Metrics used by the {@link BaseMetadataStore} and {@link TableBasedMetadataStore} classes.
  */
 public class StorageMetadataMetrics {
```
Comment (andreipaduroiu) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/StorageMetadataMetrics.java:22]: static class

File: shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java [hunk 6]
```diff
@@ -135,6 +143,16 @@
     public static final String STORAGE_METADATA_TABLE_GET_LATENCY = PREFIX + "segmentstore.storage.metadata_table_get_latency_ms";      // Histogram
     public static final String STORAGE_METADATA_TABLE_WRITE_LATENCY = PREFIX + "segmentstore.storage.metadata_table_write_latency_ms";  // Histogram
 
+    public static final String STORAGE_METADATA_FOUND_IN_TXN = PREFIX + "segmentstore.storage.metadata_found_in_txn";       // Counter
+    public static final String STORAGE_METADATA_FOUND_IN_BUFFER = PREFIX + "segmentstore.storage.metadata_found_in_buffer"; // Counter
+    public static final String STORAGE_METADATA_FOUND_IN_CACHE = PREFIX + "segmentstore.storage.metadata_found_in_cache";   // Counter
```
Comment (andreipaduroiu) [inline shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java:148]: Should we consider renaming these to follow more standard nomenclatures, such as "miss" or "hits"?

File: shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java [hunk 7]
```diff
@@ -135,6 +143,16 @@
...
+    public static final String STORAGE_METADATA_FOUND_IN_STORE = PREFIX + "segmentstore.storage.metadata_found_in_store";   // Counter
+    public static final String STORAGE_METADATA_NOT_FOUND = PREFIX + "segmentstore.storage.metadata_not_found";             // Counter
+
+    public static final String STORAGE_METADATA_BUFFER_SIZE = PREFIX + "segmentstore.storage.metadata_buffer_size";         // Gauge
```
Comment (andreipaduroiu) [inline shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java:152]: Always include the unit of measure for sizes. In this case, maybe add a `_bytes` suffix to all these sizes.

**Review by andreipaduroiu — CHANGES_REQUESTED** [review]

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentStoreMetrics.java (hunk 1 above)
Comment (sachin-j-joshi) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentStoreMetrics.java:105]: good catch. `this.storageQueueSize.close();` was missing

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageMetrics.java (hunk 2 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageMetrics.java:22]: in java `static` has different meaning than C#.
In C# it mean you can not have instance variable. 

In java you can have nested classes that have reference to _enclosing outer object_. When you have static in java it means don't include ref to outer class. By this definition all top level classes are static already.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/StorageMetadataMetrics.java (hunk 5 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/StorageMetadataMetrics.java:22]: same as above.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/StatsReporter.java (hunk 4 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/StatsReporter.java:16]: fixed.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java (hunk 3 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java]: fixed

File: shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java (hunk 6 above)
Comment (sachin-j-joshi) [reply shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java]: done

File: shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java (hunk 7 above)
Comment (sachin-j-joshi) [reply shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java]: changed names.

**Review by andreipaduroiu — APPROVED** [review]

**Pull request merged**